### ISSUE TITLE
Mask mode tests

### ISSUE THREAD

**Issue by magnumripper** [magnumripper/jtrTestSuite#23]
See https://github.com/magnumripper/JohnTheRipper/issues/841 for lots of very small tests that TS should verify (if there is a mask mode present at all). They will ensure ranges and placeholders are parsed (or not) correctly, when things are escaped.

**Comment by jfoug** [comment]
Should this be an external perl script (like the dyna-compiler test suite fuzzer???)

I think an external would be better than incorporated into the current TS.  What we 'could' do, is to spawn these external script, from within the TS script, without having to 'overload' the current script, with additional complexity, doing things that are quite different than all current testing.

**Comment by magnumripper** [comment]
Yes I think so. Actually I think the priority is to establish *manual* tests that should be made. When that is done, we almost don't need them automated (testing only needed when changing mask mode) but at the same time it would be easy to do so.

**Comment by jfoug** [comment]
But placing them into a script is not  a bad idea.  It provides a simple regression test, to make sure changes do not break something.   It is really all that the test suite is.  Just a regression tool, once all formats are in it.

**Comment by magnumripper** [comment]
True. Also, I really liked the design of the resume tests. One generic `doOneRestore()` function, and a number of calls to it. That ended up really well. We could do something similar, perhaps even come up with a single generic function that can test most modes.

**Comment by jfoug** [comment]
The whole TS is like that.  There is a 'core' function, and we simply call that, based upon the *.dat file, and user input whittling that data down.   That core function may be a 'loop',

**Comment by jfoug** [comment]
See #75   I have a plan on testing the rules, and I think I can do something similar to test mask mode.  I will know more once I get the rules tester working. Right now, it is just an idea that popped into my head on a way to do it ;)
